pkg/authentication/externalidp/aws_client.go

```diff
@@ -0,0 +1,203 @@
+	//nolint:gosec
+	AWSAuthCredentialKey = "X-Amz-Credential"
+	AWSAuthDateKey       = "X-Amz-Date"
+	AWSAuthExpiresKey    = "X-Amz-Expires"
+	//nolint:gosec
+	AWSAuthSecurityTokenKey  = "X-Amz-Security-Token"
+	AWSAuthSignedHeadersKey  = "X-Amz-SignedHeaders"
+	AWSAuthSignatureKey      = "X-Amz-Signature"
+	AWSDatetimeFormat        = "20060102T150405Z"
+	AWSCredentialTimeFormat  = "20060102"
+	AWSDefaultSTSLoginExpire = 15 * time.Minute
+)
+
+var ErrAWSCredentialsExpired = errors.New("AWS credentials expired")
+var ErrRetrievingToken = errors.New("failed to retrieve token")
+
+type LoginResponse struct {
+	Token string
+}
+
+type IDPProvider interface {
+	Login() (LoginResponse, error)
+}
+
+type AWSProvider struct {
+	params         AWSIAMParams
+	serverEndpoint string
+	client         *http.Client
+}
+
+type IdentityTokenInfo struct {
+	Method             string   `json:"method"`
+	Host               string   `json:"host"`
+	Region             string   `json:"region"`
+	Action             string   `json:"action"`
+	Date               string   `json:"date"`
+	ExpirationDuration string   `json:"expiration_duration"`
+	AccessKeyID        string   `json:"access_key_id"`
+	Signature          string   `json:"signature"`
+	SignedHeaders      []string `json:"signed_headers"`
+	Version            string   `json:"version"`
+	Algorithm          string   `json:"algorithm"`
+	SecurityToken      string   `json:"security_token"`
+}
+
+type AWSIAMParams struct {
+	ProviderType        string
+	TokenRequestHeaders map[string]string
+	URLPresignTTL       time.Duration
+	TokenTTL            time.Duration
+}
+
+func NewAWSProvider(params AWSIAMParams, serverEndpoint string, httpClient *http.Client) *AWSProvider {
+	return &AWSProvider{
+		params:         params,
+		serverEndpoint: serverEndpoint,
+		client:         httpClient,
+	}
+}
+
+func (p *AWSProvider) Login() (LoginResponse, error) {
+	jwt, err := getJWT(&p.params, p.serverEndpoint, p.client)
+	resp := LoginResponse{Token: jwt}
+	return resp, err
+}
+
+func getJWT(params *AWSIAMParams, serverEndpoint string, httpClient *http.Client) (string, error) {
+	ctx := context.TODO()
+	identityToken, err := getIdentityToken(ctx, params)
+	if err != nil {
+		return "", err
+	}
+
+	client, err := apigen.NewClientWithResponses(
+		serverEndpoint,
+		apigen.WithHTTPClient(httpClient),
+	)
+	if err != nil {
+		return "", err
+	}
+
+	tokenTTL := int(params.TokenTTL.Seconds())
+	externalLoginInfo := apigen.ExternalLoginInformation{
+		IdentityRequest: map[string]interface{}{
+			"identity_token": identityToken,
+		},
+		TokenExpirationDuration: &tokenTTL,
+	}
+	externalPrincipalLoginResp, err := client.ExternalPrincipalLoginWithResponse(ctx, apigen.ExternalPrincipalLoginJSONRequestBody(externalLoginInfo))
+	if err != nil {
+		return "", err
+	}
+	if externalPrincipalLoginResp == nil || externalPrincipalLoginResp.JSON200 == nil {
+		return "", ErrRetrievingToken
+	}
+	return externalPrincipalLoginResp.JSON200.Token, nil
```

Comment:
1. Check http issue after regular error with `ResponseAsError` 
2. Return struct AuthenticationToken not just the token, it contains more info and allows extending
3. No need for those `nil` checks below

```go
import "github.com/treeverse/lakefs/pkg/api/helpers"

/// 
if err != nil {
    ... 
}
err = helpers.ResponseAsError(res)
if err != nil {
	return nil, err
}
return res.JSON200
```